src/cascadia/WindowsTerminal/AppHost.cpp

```diff
@@ -78,40 +78,189 @@ AppHost::~AppHost()
 // - <none>
 void AppHost::_HandleCommandlineArgs()
 {
-    if (auto commandline{ GetCommandLineW() })
+    const auto args{ _NormalizedArgs() };
+    if (!args.empty())
     {
-        int argc = 0;
+        const auto result = _logic.SetStartupCommandline({ args });
+        const auto message = _logic.EarlyExitMessage();
+        if (!message.empty())
+        {
+            const auto displayHelp = result == 0;
+            const auto messageTitle = displayHelp ? IDS_HELP_DIALOG_TITLE : IDS_ERROR_DIALOG_TITLE;
+            const auto messageIcon = displayHelp ? MB_ICONWARNING : MB_ICONERROR;
+            // TODO:GH#4134: polish this dialog more, to make the text more
+            // like msiexec /?
+            MessageBoxW(nullptr,
+                        message.data(),
+                        GetStringResource(messageTitle).data(),
+                        MB_OK | messageIcon);
+
+            ExitProcess(result);
+        }
+    }
+}
 
-        // Get the argv, and turn them into a hstring array to pass to the app.
-        wil::unique_any<LPWSTR*, decltype(&::LocalFree), ::LocalFree> argv{ CommandLineToArgvW(commandline, &argc) };
-        if (argv)
+// Method Description:
+// - Retrieve the command line arguments passed,
+//   prepend the full name of the application in case it was missing (GH#4170),
+//   and return the arguments as vector of hstrings.
+// Arguments:
+// - <none>
+// Return Value:
+// - Program arguments as vector of hstrings. If the function fails it returns an empty vector.
+std::vector<winrt::hstring> AppHost::_NormalizedArgs() const noexcept
+{
+    try
+    {
+        // get the full name of the terminal app
+        auto hProc{ GetCurrentProcess() };
+        DWORD appPathLen{};
+        DWORD capacity{ 256UL }; // begin with a reasonable string size to avoid any additional iteration of the do-while loop
+        std::wstring appPath{};
+        do
         {
-            std::vector<winrt::hstring> args;
-            for (auto& elem : wil::make_range(argv.get(), argc))
+            if (capacity > (std::numeric_limits<DWORD>::max() >> 1))
             {
-                args.emplace_back(elem);
+                return {};
             }
 
-            const auto result = _logic.SetStartupCommandline({ args });
-            const auto message = _logic.EarlyExitMessage();
-            if (!message.empty())
+            capacity <<= 1;
+            appPath.resize(static_cast<size_t>(capacity));
+            appPathLen = GetModuleFileNameExW(hProc, nullptr, appPath.data(), capacity);
+        } while (appPathLen >= capacity - 1UL);

```

Comment:
For this one, we can just use `wil::GetModuleFilenameW<std::wstring>(nullptr)` :smile:

Reply:
it does the process handle management and string resizing and everything